main/guides/getting-started/contract-rpc.md

```diff
@@ -0,0 +1,410 @@
+
+:::
+
+## Specifying Offers
+
+Recall that for an agent within the JavaScript VM,
+[E(zoe).offer(...)](../../reference/zoe-api/zoe.md#e-zoe-offer-invitation-proposal-paymentkeywordrecord-offerargs) takes an `Invitation` and optionally a `Proposal` with `{ give, want }`, a `PaymentKeywordRecord`, and `offerArgs`; it returns a `UserSeat` from which we can [getPayouts()](../../reference/zoe-api/user-seat.md#e-userseat-getpayouts).
+
+![Zoe API diagram, simplified](./assets/zoe-simp.svg)
+
+In the Smart Wallet architecture, a client uses an `OfferSpec` to
+tell its `SmartWallet` how to conduct an offer.
+It includes an `invitationSpec` to say what invitation to pass to Zoe. For example:
+
+<<< @/snippets/test-marshal.js#exInvitationSpec
+
+Here the `SmartWallet` calls `E(zoe).getPublicFacet(instance)` and then
+uses the `publicInvitationMaker` and `invitationArgs` to call the contract's
+public facet.
+
+![InvitationSpec sequence diagram](./assets/inv-spec.svg)
+
+<!-- TODO: SVG diagram in /assets/ -->
+
+::: tip InvitationSpec Patterns
+For more `InvitationSpec` examples, see [How to make an offer from a dapp via the smart wallet? \(InvitationSpec Patterns\) · #8082](https://github.com/Agoric/agoric-sdk/discussions/8082) July 2023
+:::
+
+The client fills in the proposal, which instructs the `SmartWallet`
+to withdraw corresponding payments to send to Zoe.
+
+<<< @/snippets/test-marshal.js#exBridgeAction
+
+But recall the `spend_action` field in `MsgWalletSpendAction` is a string.
+In fact, the expected string in this case is of the form:
+
+<<< @/snippets/test-marshal.js#exSpendActionCk
+
+We recognize `"method":"executeOffer"` and such, but
+`body:`, `slots:`, and `$1.Alleged: Gold Brand` need further explanation.
+
+### Marshalling Amounts and Instances
+
+To start with, amounts include `bigint`s. The `@endo/marshal` API handles those:
+
+<<< @/snippets/test-marshal.js#marshal-json-steroids
+
+To marshal brands and instances, recall from the discussion of [marshal in eventual send](../js-programming/eventual-send.md#e-and-marshal-a-closer-look)
+how remotables are marshalled with a translation table.
+
+The [Agoric Board](../../reference/repl/board.md) is a well-known name service that issues
+plain string identifiers for object identities and other passable _keys_ (that is: passable values excluding promises and errors).
+Contracts and other services can use its table of identifiers as a marshal
+translation table:
+
+<<< @/snippets/test-marshal.js#boardMarshal
+
+To reverse the process, clients can mirror the on-chain board translation
+table by synthesizing a remotable for each reference marker received:
+
+<<< @/snippets/test-marshal.js#makeBoardContext
+
+Now we can take results of vstorage queries for `Data('published.agoricNames.brand')` and `Data('published.agoricNames.instance')` unmarshal ("ingest") them:
```

Comment:
"ingest" is probably an unnecessary concept. just use `fromCapData`.

(having studied popular definitions of "serialize" and "marsha" some more, I think it's not wrong to use them with the CapData structure. I regret the renaming to `toCapData` and `fromCapData`)

Reply:
>"ingest" is probably an unnecessary concept. just use fromCapData.

Agree.

> (having studied popular definitions of "serialize" and "marsha" some more, I think it's not wrong to use them with the CapData structure. I regret the renaming to toCapData and fromCapData)

Disagree. 

> In computing, serialization (or serialisation) is the process of translating a [data structure](https://en.wikipedia.org/wiki/Data_structure) or [object](https://en.wikipedia.org/wiki/Object_(computer_science)) state into a format that can be stored (e.g. [files](https://en.wikipedia.org/wiki/Computer_file) in [secondary storage devices](https://en.wikipedia.org/wiki/Secondary_storage_devices), [data buffers](https://en.wikipedia.org/wiki/Data_buffer) in primary storage devices) or transmitted (e.g. [data streams](https://en.wikipedia.org/wiki/Data_stream) over [computer networks](https://en.wikipedia.org/wiki/Computer_networks)) and reconstructed later (possibly in a different computer environment).[[1]](https://en.wikipedia.org/wiki/Serialization#cite_note-1)

The `serialize` function was not doing this. It was turning one object graph (original) into another (Cap Data structure).

Reply:
The Cap Data structure is straightforward to store / transmit.